zkSNACKs/WalletWasabi
# Tor control & graceful Tor shutdown

This is a part of #5681 and a follow-up PR to #5755.

In short, WW connects to Tor Control at the application start and the connection is supposed to be persistent in the sense that it should exist as long as WW is up and running. Once WW is closed by a user (and TerminateTorOnExit is `true`), Tor process will be gracefully shut down.

The implementation may seem a bit over the top but that's how [spec - 3.23. TAKEOWNERSHIP](https://gitweb.torproject.org/torspec.git/tree/control-spec.txt) says it should be done.

## Fixed port for Tor control

I picked the port 37151 for Tor control listener. The default port is 9051. The last two digits of the chosen port and the default port are the same to make it easier to remember the number (🙏 ).

The port 37151 does not seem to be used by existing applications based on:

* https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml?&page=131
* https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers

## Why not `--ControlPort auto`?

Tor offers `--ControlPort auto` argument that is really useful because it allows the port for Tor control to be picked automatically by Tor so we would avoid accidental port collisions for sure. It seems very easy to use the feature, however, currently, I think it is not [as straighforward as one may think](https://tor.stackexchange.com/questions/22342/how-to-correctly-read-the-control-port-when-tor-is-started-with-controlport-a). If you can answer the previous question, please do so. David [re-posted the issue on the Tor's bugtracker](https://gitlab.torproject.org/tpo/core/tor/-/issues/40386).

So this feature is reserved for a future PR, hopefully.

<details>
  <summary>Details</summary>

There are probably two methods one can add this "random Tor control port" feature. However, I find one thing about this feature quite off-putting: WW needs to store the tor control port somewhere so that when WW is started again it uses this port (for the case when Tor is not teminated when WW terminates).

The first method is not exactly trivial and the second is actually not good for the scenario when Tor continues running after WW terminates.

### 1) Use `--ControlPort auto`

The correct procedure to use this feature [seems](https://gitlab.torproject.org/tpo/core/tor/-/issues/40386#note_2735298) to be:

1. Make sure that `<TEMPORARY_FILE_PATH>` does not exist [to eliminate risk that you read some old value].
    * Make sure that the path is somewhere randomly generated that only you have write access, to eliminate the risk of somebody else writing to it. (Using `mkstemp` or similar should take care of that part.) 
1. Start Tor process with the arguments
1. Wait until `<TEMPORARY_FILE_PATH>` file exists [the file may exist but its content may not be flushed just yet now]
    * Monitor the Tor process that you launched, in case it exits for some reason before it can open this port.
1. Repeat
    * Open `<TEMPORARY_FILE_PATH>` file for reading and read a single line (ending with `\n`)
    * When the line starts with `PORT=`, you can parse the port -> success & break from loop.

### 2) Use `__OwningControllerFD`

Blog post: https://blog.torproject.org/tor-0336-released-new-stable-series 

> Tor can now start with a preauthenticated control connection created by the process that launched it. This feature is meant for use by programs that want to launch and manage a Tor process without allowing other programs to manage it as well. For more information, see the `__OwningControllerFD` option documented in control-spec.txt. Closes ticket [23900](https://gitlab.torproject.org/legacy/trac/-/issues/23900).

Idea is to pass `__OwningControllerFD` as an uint64_t number. 

### Conclusion

Implementing a random control port feature seems to me like a non-trivial programming task and if the feature is not really stable it would be very hard to provide support for the feature. The benefit of the feature seems quite small compared to the mentioned concerns. What do you think?

### Resources:

* Source code
    * PR adding `OwningControllerFD`: https://github.com/torproject/tor/commit/f1bf9bf8198fcfaf078fdc12eb2ad5adf1901d29
        * https://github.com/torproject/tor/blob/d4718e4db9779ba11baf7ee07e093d0c9e6f6bae/src/feature/api/tor_api.c#L114
        * https://github.com/torproject/tor/blob/0cf3ab54f6701e37a28a3ad06d4dae1d7a031007/src/core/mainloop/mainloop.c#L1476
        * https://github.com/torproject/tor/blob/0cf3ab54f6701e37a28a3ad06d4dae1d7a031007/src/app/config/or_options_st.h#L549
* https://github.com/torproject/tor/commit/fe3bacf50a8d6e7fcec5f8bb1d5c0ac32c41a57c should work on Windows too
* https://github.com/dotnet/runtime/pull/34727 - "Add Socket(SafeSocketHandle) ctor" (on 11 Apr 2020)
    * https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.-ctor?view=net-5.0#System_Net_Sockets_Socket__ctor_System_Net_Sockets_SafeSocketHandle_
    * https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.safehandle.handle?view=net-5.0
* https://stackoverflow.com/questions/58855312/c-sharp-proper-use-of-dispose-with-safesockethandle - very useful

</details>


## TakeOwnership implementation

Lucas uses [__OwningControllerProcess](https://github.com/lontivero/Torino/blob/05b5bb7282750209909d768a4afee85a96b075f0/src/TorLauncher.cs#L46) argument to start Tor and make it rely on a starting OS process. However, WW allows the following scenario to happen:

1. "Terminate Tor on exit" is not enabled in WW settings.
1. WW is started but Tor cannot be started with `__OwningControllerProcess` argument.
1. User enables "Terminate Tor on exit" in WW settings but it has no effect because old Tor is still running.

This is why I think using `TAKEOWNERSHIP` command is useful here.

## Testing

I would be grateful for any help with testing. One risk I can see is that TCP connection between WW and Tor control will time out after some significant amount of time. I'm not sure what the best solution is for this or if it can happen really (I believe a typical solution for this is to send a ping request through the TCP connection.)

### Testing instructions

1) One should start WW and check that Tor process is terminated when `TerminateTorOnExit=true` (the setting in WW settings) and that it is not when `TerminateTorOnExit=false`. 

2) For `TerminateTorOnExit=true` case, one should verify that when you kill WW process (e.g. using Windows Task Manager), Tor is terminated too (logs should confirm that the termination was graceful).

3) It's useful to test another case when `TerminateTorOnExit=true` and you just let WW open over night whether Tor will be running in the morning or not - it should be running.

(cc @yahiheb @MaxHillebrand)

PS: Yahia found one bug :tada:
